### PR TITLE
Enable OAuth PKCE Extension

### LINKED ISSUE #21913: Support OAuth PKCE to allow third-party single-page apps to utilize best practice when authorizing with an instance
### Pitch

Mastodon [hasn't configured PKCE on Doorkeepr](https://github.com/mastodon/mastodon/discussions/17207?sort=top#discussioncomment-1913479). Will it be possible to make that OAuth extension accessible?

Also, if PKCE is used, can the client secret _not_ be shared by the instance when the app registers with it?

### Motivation

If you are to write a third-party app that runs completely off the browser and it needs to authorize with an OAuth capable server, the best practice is to use ["the Authorization Code flow with the PKCE extension"](https://oauth.net/2/browser-based-apps/). Since Mastodon doesn't offer that extension, such a third-party client would have to store [the client secret in `localStorage`](https://github.com/takahashim/mastodon-access-token/blob/master/index.html#L189) in order to retrieve access tokens. Storing the client secrets in `localStorage` is [insecure](https://thesecurityvault.com/auth-token-in-localstorage/).

Why not create a backend to keep the client secret safe? Well now you have [to store the client secret for each mastodon instance in a database](https://github.com/diddledani/twitodon/blob/fd943cc163f516b4e943c529ae77f5a17f731407/api/mastodon.mjs#L39). Also, the creation of the backend is extra effort, complexity, and resources 💸.

### PR DESCRIPTION
**Pull request by ThisIsMissEm**
This pull request partially supersedes #30329, in that I've separated out the PKCE changes from the other changes.

As per [OAuth 2.0 Security Best Current Practices](https://www.ietf.org/archive/id/draft-ietf-oauth-security-topics-27.html#name-authorization-code-grant), we MUST support PKCE.

Whilst PKCE is recommended for all applications to use, forcing PKCE usage is a breaking change, so for now we do not force PKCE usage, only support it. PKCE is an important security mechanism for preventing [OAuth Authorization Code Injection Attacks](https://www.ietf.org/archive/id/draft-ietf-oauth-security-topics-27.html#name-pkce). **Note:** OAuth 2.1 requires the usage of PKCE.

In accordance with best practices, we only support S256, and explicitly do not support plain for PKCE `code_challenge_method`. Unfortunately, Doorkeeper [does not support](https://github.com/doorkeeper-gem/doorkeeper/issues/1717) a setting for which `code_challenge_method`s are supported, hence adding a `OauthPreAuthorizationExtension` module and changing the `doorkeeper.errors.messages.invalid_code_challenge_method` value.

I've also attempted to add some better tests to ensure certain elements are definitely present on the page, such that we're not just relying on the `response.body` containing a given string, but actually that the element exists.

Fixes: #21913

### CODE REVIEW

File: spec/requests/well_known/oauth_metadata_spec.rb [hunk 1]
```diff
@@ -29,7 +29,10 @@
       revocation_endpoint: oauth_revoke_url(protocol: protocol),
       scopes_supported: Doorkeeper.configuration.scopes.map(&:to_s),
       response_types_supported: Doorkeeper.configuration.authorization_response_types,
+      response_modes_supported: Doorkeeper.configuration.authorization_response_flows.flat_map(&:response_mode_matches).uniq,
+      token_endpoint_auth_methods_supported: %w(client_secret_basic client_secret_post),
```
Comment (ThisIsMissEm) [inline spec/requests/well_known/oauth_metadata_spec.rb:33]: For some reason #29191 didn't actually include tests for these, despite them being present.

File: config/application.rb [hunk 2]
```diff
@@ -118,6 +118,7 @@ class Application < Rails::Application
       Doorkeeper::Application.include ApplicationExtension
       Doorkeeper::AccessGrant.include AccessGrantExtension
       Doorkeeper::AccessToken.include AccessTokenExtension
+      Doorkeeper::OAuth::PreAuthorization.include OauthPreAuthorizationExtension
```
Comment (ThisIsMissEm) [inline config/application.rb:121]: I'm honestly not sure I've done this in the "correct" way, but it's temporary until https://github.com/doorkeeper-gem/doorkeeper/issues/1717 is resolved.

File: app/lib/oauth_pre_authorization_extension.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+module OauthPreAuthorizationExtension
+  extend ActiveSupport::Concern
+
+  included do
+    validate :code_challenge_method_s256, error: Doorkeeper::Errors::InvalidCodeChallengeMethod
+  end
+
+  def validate_code_challenge_method_s256
+    code_challenge.blank? || (code_challenge_method.present? && code_challenge_method == 'S256')
```
Comment (ThisIsMissEm) [inline app/lib/oauth_pre_authorization_extension.rb:11]: I think we can just write this as:

```suggestion
    code_challenge.blank? || code_challenge_method == 'S256'
```

Comment (oneiros) [reply app/lib/oauth_pre_authorization_extension.rb:11]: Yes, I agree.

File: config/application.rb (hunk 2 above)
Comment (oneiros) [reply config/application.rb:121]: Given that, I believe it is totally fine for now.

File: spec/controllers/oauth/authorizations_controller_spec.rb [hunk 4]
```diff
@@ -61,7 +78,110 @@
...
+
+            authorize_button = body.at_css('.oauth-prompt button[type="submit"]')
+
+            expect(authorize_button).to_not be_present
+          end
+
+          it 'includes an error message' do
+            subject
+            expect(error_message).to match I18n.t('doorkeeper.errors.messages.invalid_code_challenge_method')
+          end
```
Comment (oneiros) [inline spec/controllers/oauth/authorizations_controller_spec.rb:184]: All those new examples here seem to test the user interface and as such do not belong in a "controller spec".

There are two alternative options:

* If you just want to make sure certain things are present in the html template, then a "view spec" would be the correct place. Additionally you _could_ add a test here that makes sure the controller actually renders that template.
* If you want to make sure that a certain kind of client request leads to the UI being in a certain state, this should be a "system spec". If you go down that route, you might want to have these be "driven_by" `:rack_test` which is a lot faster than spinning up a browser for this (see https://rspec.info/features/6-0/rspec-rails/system-specs/system-specs/)

**Review by oneiros — CHANGES_REQUESTED** [review]
I have one remark about spec types, but otherwise this is already looking very good. Thanks for your help with this!

File: spec/controllers/oauth/authorizations_controller_spec.rb (hunk 4 above)
Comment (ThisIsMissEm) [reply spec/controllers/oauth/authorizations_controller_spec.rb:184]: I do agree, though as we were already doing the controller spec like this (assertions on response body) I'd like to run with this as-is, and create a follow up issue to address controller vs request vs system tests for OAuth flows

Comment (ThisIsMissEm) [reply spec/controllers/oauth/authorizations_controller_spec.rb:184]: I feel like system specs are probably most relevant for OAuth flows, not controller or request or view specs

Maybe we'd have a controller spec that just asserts out custom views for doorkeeper get rendered?

(also, I try really hard to limit scope of changes due to frequently being asked to break PRs down into smaller chunks, which takes significant time)

Comment (oneiros) [reply spec/controllers/oauth/authorizations_controller_spec.rb:184]: Just to be super-clear: I am not suggesting creating generic oauth flow system specs. Just moving these here out to either a view or a (single) system spec file. This would not widen the scope of this PR, just make better use of the different kinds of spec types `rspec-rails` provides.

Comment (ThisIsMissEm) [reply spec/controllers/oauth/authorizations_controller_spec.rb:184]: I'm going to wait for @ClearlyClaire's input here, since I don't want to risk overcomplicating this PR by rewriting a known working test too heavily (as it is I only made the changes I did because the assertions were quiet weak & could lead to tests passing when they shouldn't

**Comment by ClearlyClaire** [comment]
First, to make sure I understand exactly the context, as I'm not so familiar with Doorkeepr and OAuth:
- PKCE is an extension of OAuth made so that the OAuth provider can ensure that the application that received the code is truly the one that requested it, before giving it the authentication token
- this is primarily meant for public clients, which Mastodon doesn't really support at the moment (which is a whole other issue), but can still be useful for confidential clients in some cases, and has no real drawback, and therefore is a MUST in the latest OAuth 2.0 best practices
- Doorkeeper does support it natively, provided the database has the correct columns and the parameters are passed properly
- there is some discussion regarding PKCE and refresh tokens (which we don't currently support either) but my understanding is that for the most part, those are orthogonal and refresh tokens do not typically involve a potentially unsecure channel, unlike the initial authentication

Please correct me if I'm wrong on any on this.

(Now I'll look at the code itself and especially where you tagged me)

File: spec/controllers/oauth/authorizations_controller_spec.rb (hunk 4 above)
Comment (ClearlyClaire) [reply spec/controllers/oauth/authorizations_controller_spec.rb:184]: I agree with @oneiros that this kind of tests belong in system tests. I'd suggest looking at the existing `spec/system/oauth_spec.rb` tests to see if those could be moved there instead without a lot more effort, but if that amount of work is prohibitive at the moment, I'm ok with the consolidation happening in another PR (note that I have not actually reviewed the specs yet, just quickly read through them, and I'm commenting on the general approach)

**Comment by ThisIsMissEm** [comment]
@ClearlyClaire 
> this is primarily meant for public clients, which Mastodon doesn't really support at the moment (which is a whole other issue), but can still be useful for confidential clients in some cases, and has no real drawback, and therefore is a MUST in the latest OAuth 2.0 best practices

It was originally for public clients, but is now a MUST for all clients due to authorization code injection attacks.

> there is some discussion regarding PKCE and refresh tokens (which we don't currently support either) but my understanding is that for the most part, those are orthogonal and refresh tokens do not typically involve a potentially unsecure channel, unlike the initial authentication

tl;dr; let's disregard refresh tokens for now, and focus on PKCE. (#27948 has been made a draft to indicate it should **not** be in 4.3)

Doorkeeper's refresh token model is fundamentally broken though, because it doesn't keep trust chains (basically if you revoke the access token the refresh token can remain valid, and if you do a `access token -> refresh token -> access token -> refresh token` then there's something where the revocation of the original access token doesn't result in the second token set being revoked or something. (I'm still trying to get my head fully around this.

From Auth0:
- Refresh token rotation guarantees that every time an application exchanges a refresh token to get a new access token, a new refresh token is also returned.
- [Automatic Reuse Detection for Refresh Tokens](https://auth0.com/blog/refresh-tokens-what-are-they-and-when-to-use-them/#Refresh-Token-Automatic-Reuse-Detection)

tl;dr; is that you should actually have a separate oauth_refresh_tokens table, instead of just a column in the oauth_access_tokens table, as you need to be able to follow the chain of tokens.

File: spec/controllers/oauth/authorizations_controller_spec.rb (hunk 4 above)
Comment (ThisIsMissEm) [reply spec/controllers/oauth/authorizations_controller_spec.rb:184]: Yeah, I'd like to do a review of the oauth specs as a separate task after implement PKCE. Only reason I noticed was because they were already doing stuff like

```
        expect(response.body).to match(/Authorize/)
```

Which isn't great at all, because the string Authorize could appear *anywhere* on the response body.

Comment (ThisIsMissEm) [reply spec/controllers/oauth/authorizations_controller_spec.rb:184]: Have created https://github.com/mastodon/mastodon/issues/31150 to track the OAuth Specs state.

**Comment by ThisIsMissEm** [comment]
@ClearlyClaire @oneiros also, on "public clients", I'm not sure if it actually makes long-term sense to continue support for `POST /api/v1/apps` once we have implemented [OAuth Client ID Metadata Documents](https://drafts.aaronpk.com/draft-parecki-oauth-client-id-metadata-document/draft-parecki-oauth-client-id-metadata-document.html), since these are superior to DCR for federated many clients to many servers setups.

(Note: Client ID Metadata Documents are currently an I-D and have been put in front of the IETF OAuth working group at IETF 120 this week, where overall they were well received.

**Comment by ThisIsMissEm** [comment]
Added an issue about Client ID Metadata Documents here: https://github.com/mastodon/mastodon/issues/31151

**Review by ClearlyClaire — APPROVED** [review]

**Review by oneiros — APPROVED** [review]
